[SPMD][BE] Split graph optimization module into multiple submodules

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #101012

Graph optimization module is becoming too large for efficient development. This PR splits it into smaller ones. This PR also changes the naming of graph optimizations as not all passes target the performance.

Differential Revision: [D45703835](https://our.internmc.facebook.com/intern/diff/D45703835/)

fegin committed:
[SPMD][BE] Split graph optimization module into multiple submodules

Graph optimization module is becoming too large for efficient development. This PR splits it into smaller ones. This PR also changes the naming of graph optimizations as not all passes target the performance.

Differential Revision: [D45703835](https://our.internmc.facebook.com/intern/diff/D45703835/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD][BE] Split graph optimization module into multiple submodules"


Graph optimization module is becoming too large for efficient development. This PR splits it into smaller ones. This PR also changes the naming of graph optimizations as not all passes target the performance.

Differential Revision: [D45703835](https://our.internmc.facebook.com/intern/diff/D45703835/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD][BE] Split graph optimization module into multiple submodules"


Graph optimization module is becoming too large for efficient development. This PR splits it into smaller ones. This PR also changes the naming of graph optimizations as not all passes target the performance.

Differential Revision: [D45703835](https://our.internmc.facebook.com/intern/diff/D45703835/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD][BE] Split graph optimization module into multiple submodules"


Graph optimization module is becoming too large for efficient development. This PR splits it into smaller ones. This PR also changes the naming of graph optimizations as not all passes target the performance.

Differential Revision: [D45703835](https://our.internmc.facebook.com/intern/diff/D45703835/)

[ghstack-poisoned]

lessw2020 commented on torch/distributed/_spmd/graph_transformations/common.py:
```diff
@@ -0,0 +1,288 @@
+            _optimized_func.add(func_key)
+
+            prefix = f"after_{func.__name__}"
+            if _dump_graph_folder:
```
+1 on having the option to dump graphs between passes for review.

lessw2020 commented on torch/distributed/_spmd/graph_transformations/optimizer.py:
```diff
@@ -0,0 +1,461 @@
+)
+def remove_copy_from_optimizer(gm: IterGraphModule) -> None:
+    """
+    Erase the the orphant copy_ that generated when tracing optimizer.
```
minor nit but "orphaned"

lessw2020 commented on torch/distributed/_spmd/graph_transformations/optimizer.py:
```diff
@@ -0,0 +1,461 @@
+                nodes.append(None)
+            continue
+        elif node.op == OP.CALL_FUNCTION and str(node.target).startswith(
+            "aten._foreach_add"
```
very minor, but maybe move "aten._foreach_add" from hardcoded reference here, to a variable/const alongside max_step_distance above, for visibility/future changes.

lessw2020 reviewed (approved):
looks good!

fegin committed:
Update on "[SPMD][BE] Split graph optimization module into multiple submodules"


Graph optimization module is becoming too large for efficient development. This PR splits it into smaller ones. This PR also changes the naming of graph optimizations as not all passes target the performance.

Differential Revision: [D45703835](https://our.internmc.facebook.com/intern/diff/D45703835/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD][BE] Split graph optimization module into multiple submodules"


Graph optimization module is becoming too large for efficient development. This PR splits it into smaller ones. This PR also changes the naming of graph optimizations as not all passes target the performance.

Differential Revision: [D45703835](https://our.internmc.facebook.com/intern/diff/D45703835/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD][BE] Split graph optimization module into multiple submodules"


Graph optimization module is becoming too large for efficient development. This PR splits it into smaller ones. This PR also changes the naming of graph optimizations as not all passes target the performance.

Differential Revision: [D45703835](https://our.internmc.facebook.com/intern/diff/D45703835/)

[ghstack-poisoned]
